transactions/src/main/java/org/jboss/as/txn/service/internal/tsr/JCAOrderedLastSynchronizationList.java

```diff
@@ -0,0 +1,146 @@
+        Map<Transaction, JCAOrderedLastSynchronizationList> jcaOrderedLastSynchronizations) {
+        this.tx = tx;
+        this.jcaOrderedLastSynchronizations = jcaOrderedLastSynchronizations;
+    }
+
+    public void addInterposedSynchronization(Synchronization synchronization) {
+        if (synchronization.getClass().getName().startsWith("org.jboss.jca")) {
+            if (TransactionLogger.ROOT_LOGGER.isTraceEnabled()) {
+                TransactionLogger.ROOT_LOGGER.trace("JCAOrderedLastSynchronizationList.jcaSyncs.add - Class: " + synchronization.getClass() + " HashCode: "
+                    + synchronization.hashCode() + " toString: " + synchronization);
+            }
+            jcaSyncs.add(synchronization);
+        } else {
+            if (TransactionLogger.ROOT_LOGGER.isTraceEnabled()) {
+                TransactionLogger.ROOT_LOGGER.trace("JCAOrderedLastSynchronizationList.preJcaSyncs.add - Class: " + synchronization.getClass() + " HashCode: "
+                    + synchronization.hashCode() + " toString: " + synchronization);
+            }
+            preJcaSyncs.add(synchronization);
+        }
+    }
+
+    @Override
+    public void beforeCompletion() {
+        for (Synchronization preJcaSync : preJcaSyncs) {
+            if (TransactionLogger.ROOT_LOGGER.isTraceEnabled()) {
+                TransactionLogger.ROOT_LOGGER.trace("JCAOrderedLastSynchronizationList.preJcaSyncs.before_completion - Class: " + preJcaSync.getClass() + " HashCode: "
+                    + preJcaSync.hashCode()
+                    + " toString: "
+                    + preJcaSync);
+            }
+            preJcaSync.beforeCompletion();
+        }
+        for (Synchronization jcaSync : jcaSyncs) {
+            if (TransactionLogger.ROOT_LOGGER.isTraceEnabled()) {
+                TransactionLogger.ROOT_LOGGER.trace("JCAOrderedLastSynchronizationList.jcaSyncs.before_completion - Class: " + jcaSync.getClass() + " HashCode: "
+                    + jcaSync.hashCode()
+                    + " toString: "
+                    + jcaSync);
+            }
+            jcaSync.beforeCompletion();
+        }
+    }
+
+    @Override
+    public void afterCompletion(int status) {
+        // The list should be iterated in reverse order - has issues with EJB3 if not
+        // https://github.com/jbosstm/narayana/blob/master/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TwoPhaseCoordinator.java#L509
+        Collections.reverse(preJcaSyncs);
+        for (Synchronization preJcaSync : preJcaSyncs) {
+            if (TransactionLogger.ROOT_LOGGER.isTraceEnabled()) {
+                TransactionLogger.ROOT_LOGGER.trace("JCAOrderedLastSynchronizationList.preJcaSyncs.afterCompletion - Class: " + preJcaSync.getClass() + " HashCode: "
+                    + preJcaSync.hashCode()
+                    + " toString: " + preJcaSync);
+            }
+            preJcaSync.afterCompletion(status);
```

Comment:
What happens if one of these throws an exception? Is it ok for the rest of the afterCompletion calls to not run? 

I'm not sure how the transaction manager handles this. I know we use these calls internally to release EJB locks and some other cleanup, so this does not seem safe to me.